coredns/plugin/weighted.go

```diff
@@ -0,0 +1,102 @@
+package kuadrant
+
+import (
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+// Logic taken from https://github.com/coredns/coredns/blob/master/plugin/loadbalance/weighted.go
+
+type weightedRR struct {
+	dns.RR
+	weight int64
+}
+
+type weightedRRSet struct {
+	rrs []weightedRR
+	randomGen
+	mutex sync.Mutex
+}
+
+func newWeightedRRSet(rrs []weightedRR) *weightedRRSet {
+	wRRSet := &weightedRRSet{
+		rrs:       rrs,
+		randomGen: &randomUint{},
+	}
+	wRRSet.randomGen.randInit()
+	return wRRSet
+}
+
+// Random uint generator
+type randomGen interface {
+	randInit()
+	randUint(limit uint) uint
+}
+
+// Random uint generator
+type randomUint struct {
+	rn *rand.Rand
+}
+
+func (r *randomUint) randInit() {
+	r.rn = rand.New(rand.NewSource(time.Now().UnixNano()))
+}
+
+func (r *randomUint) randUint(limit uint) uint {
+	return uint(r.rn.Intn(int(limit)))
+}
+
+// Move the next expected address to the first position in the result list
+func (w *weightedRRSet) setTopRecord(address []weightedRR) {
+	itop := w.topAddressIndex(address)
+
+	if itop < 0 {
+		// internal error
+		return
+	}
+
+	if itop != 0 {
+		// swap the selected top entry with the actual one
+		address[0], address[itop] = address[itop], address[0]
+	}
+}
+
+// Compute the top (first) address index
+func (w *weightedRRSet) topAddressIndex(address []weightedRR) int {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	// Determine the weight value for each address in the answer
+	var wsum uint
+
+	for _, r := range address {
+		wsum += uint(r.weight)
+	}
+
+	// Select the first (top) IP
+	sort.Slice(address, func(i, j int) bool {
+		return address[i].weight > address[j].weight
+	})
+	v := w.randUint(wsum)
```

Comment:
So generates a random number between 0 and the sum of all the weights. Then orders the addresses by weight desc and returns the first result that has a weight or combined weight greater than the random value. So if you have a weight of 200 and another of 100 you end up with 300 and a random number between 0 and 300 picked so you are twice as likely to get one < 200 than you are to get one between >= 200 <= 300. So kinda like a weighted coin flip it wont be super predictable but will give you the weighting. At least this is how I read it

Reply:
Created follow on issue to look at improving this https://github.com/Kuadrant/dns-operator/issues/410